### ISSUE TITLE
inference 

### ISSUE THREAD

**Issue by alicera** [ultralytics/yolov5#2852]
## ❔Question
If I train a model for size 640x640.(w,h).
Then, I input a image (320x200)(w,h).

For the detect.py, how does it process the image?

**Comment by wudashuo** [comment]
https://github.com/ultralytics/yolov5/blob/238583b7d5c19029920d56c417c406c829569c75/utils/datasets.py#L122-L199

**Comment by glenn-jocher** [comment]
@alicera image sizes are independent of inference size. You specify inference size for train.py and test.py identically using `--img-size`. If no argument is passed for `--img-size` the default value of 640 is used.

```
python train.py --img 640
python detect.py --img 640
```

**Comment by alicera** [comment]
I try to use the test.py --img-size 640
I only test a image (640x425), then I print the image process size. It is the 672x448.

It will resize the long side of each of your images to 640 and add a 32x padding, maintaining aspect ratio and padding the shorter side as necessary to meet stride-multiple constraints.

The image resize to 672 ,not 640. Why?
So the coco mAP follow the 672 to test?

size 640x425
![image](https://user-images.githubusercontent.com/59388989/115515105-5e50ca80-a2b7-11eb-920c-562af6ff09e2.png)
 
size 672x448
![image](https://user-images.githubusercontent.com/59388989/115515119-614bbb00-a2b7-11eb-9780-39fd1ba5f583.png)

**Comment by glenn-jocher** [comment]
@alicera 👋 Hello, thank you for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): 0.001
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): 0.6
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): True
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): None

## test.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): 0.001
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): 0.6
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): True
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): 0.5 * maximum stride

## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): 0.25 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): 0.45
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): False
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): None

## YOLOv5 PyTorch Hub

- autoShape() class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250
